core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java

```diff
@@ -114,45 +115,51 @@ public ApiDispatchConfig findConfigForUri(String uri) {
 		return results;
 	}
 
-	public synchronized void registerServiceClient(ApiListener listener) throws ListenerException {
+	public void registerServiceClient(ApiListener listener) throws ListenerException {
 		String uriPattern = listener.getCleanPattern();
 		if(uriPattern == null)
 			throw new ListenerException("uriPattern cannot be null or empty");
 
 		String method = listener.getMethod();
 
-		ApiDispatchConfig dispatchConfig = null;
-		if(patternClients.containsKey(uriPattern))
-			dispatchConfig = patternClients.get(uriPattern);
-		else
-			dispatchConfig = new ApiDispatchConfig(uriPattern);
-
-		dispatchConfig.register(method, listener);
+		synchronized(patternClients) {
+			patternClients.computeIfAbsent(uriPattern, pattern -> new ApiDispatchConfig(pattern)).register(method, listener);
+		}
 
-		patternClients.put(uriPattern, dispatchConfig);
-		log.trace("ApiServiceDispatcher successfully registered uriPattern ["+uriPattern+"] method ["+method+"]");
+		if(log.isTraceEnabled()) log.trace("ApiServiceDispatcher successfully registered uriPattern ["+uriPattern+"] method ["+method+"]");
 	}
 
-	public synchronized void unregisterServiceClient(ApiListener listener) {
+	public void unregisterServiceClient(ApiListener listener) {
 		String method = listener.getMethod();
 		String uriPattern = listener.getCleanPattern();
 		if(uriPattern == null) {
 			log.warn("uriPattern cannot be null or empty, unable to unregister ServiceClient");
 		}
 		else {
-			ApiDispatchConfig dispatchConfig = patternClients.get(uriPattern);
-			if(dispatchConfig == null) {
-				log.warn("unable to find DispatchConfig for uriPattern ["+uriPattern+"]");
-			} else {
-				dispatchConfig.destroy(method);
+			boolean success = false;
+			synchronized (patternClients) {
+				ApiDispatchConfig dispatchConfig = patternClients.get(uriPattern);
+				if(dispatchConfig != null) {
+					if(dispatchConfig.getMethods().size() == 1) {
+						patternClients.remove(uriPattern); //Remove the entire config if there's only 1 ServiceClient registered
+					} else {
+						dispatchConfig.remove(method); //Only remove the ServiceClient as there are multiple registered
+					}
+					success = true;
+				}
+			}
 
-				log.trace("ApiServiceDispatcher successfully unregistered uriPattern ["+uriPattern+"] method ["+method+"]");
+			//Remove log statements from synchronized block
```

Comment:
commentaar mag weg, of je moet het anders formuleren: `keep log statements out of synchronized block`